lib/spack/spack/test/env.py

```diff
@@ -1059,3 +1060,145 @@ def test_toolchain_definitions_are_allowed(
+    manifest = tmp_path / "spack.yaml"
+    manifest.write_text(
+        """\
+spack:
+  specs:
+    - matrix:
+      - [dt-diamond-left,  dt-diamond-right]
+      - ["%mixed-toolchain"]
+  toolchains:
+    mixed-toolchain:
+    - spec: "%[virtuals=c] llvm"
+      when: "%c"
+    - spec: "%[virtuals=cxx] llvm"
+      when: "%cxx"
+    - spec: "%[virtuals=fortran] llvm"
+      when: "%fortran"
+  concretizer:
+    unify: true
+"""
+    )
+    with ev.Environment(tmp_path) as e:
+        e.concretize()
+        for x in e.concrete_roots():
+            if x.name in ["dt-diamond-left", "dt-diamond-right"]:
+                assert x.satisfies("%[virtuals=fortran] llvm")
+
+
+def test_matrix_toolchains_apply_both(
+    tmp_path, mutable_config, mutable_mock_env_path, temporary_store, mock_packages
+):
+    """Matrix that multiplies two conflicting toolchains with
+    two root specs.
+    """
+    manifest = tmp_path / "spack.yaml"
+    manifest.write_text(
+        """\
+spack:
+  specs:
+    - matrix:
+      - [dt-diamond-left,  dt-diamond-right]
+      - ["%toolchain-1", "%toolchain-2"]
+  toolchains:
+    toolchain-1:
+    - spec: "%[virtuals=c] llvm"
+      when: "%c"
+    - spec: "%[virtuals=cxx] llvm"
+      when: "%cxx"
+    - spec: "%[virtuals=fortran] gcc"
+      when: "%fortran"
+    toolchain-2:
+    - spec: "%[virtuals=c] gcc"
+      when: "%c"
+    - spec: "%[virtuals=cxx] gcc"
+      when: "%cxx"
+    - spec: "%[virtuals=fortran] llvm"
+      when: "%fortran"
+  concretizer:
+    unify: when_possible
+"""
+    )
+    with ev.Environment(tmp_path) as e:
+        e.concretize()
+        for x in e.concrete_roots():
+            if x.name in ["dt-diamond-left", "dt-diamond-right"]:
+                if x.satisfies("%[virtuals=c] llvm"):
+                    assert x.satisfies("%[virtuals=fortran] gcc")
+                else:
+                    assert x.satisfies("%[virtuals=c] gcc")
+                    assert x.satisfies("%[virtuals=fortran] llvm")
+
+
+def test_matrix_toolchains_three_vector(
+    tmp_path, mutable_config, mutable_mock_env_path, temporary_store, mock_packages
+):
+    """Can a toolchain be applied in a matrix"""
+    manifest = tmp_path / "spack.yaml"
+    manifest.write_text(
+        """\
+spack:
+  specs:
+    - matrix:
+      - [dt-diamond-left, dt-diamond-right]
+      - [^dt-diamond-bottom@0.9]
+      - ["%mixed-toolchain"]
+  toolchains:
+    mixed-toolchain:
+    - spec: "%[virtuals=c] llvm"
+      when: "%c"
+    - spec: "%[virtuals=cxx] llvm"
+      when: "%cxx"
+    - spec: "%[virtuals=fortran] gcc"
+      when: "%fortran"
+  concretizer:
+    unify: true
+"""
+    )
+    with ev.Environment(tmp_path) as e:
+        e.concretize()
+        for x in e.concrete_roots():
+            if x.name in ["dt-diamond-left", "dt-diamond-right"]:
+                assert x.satisfies("%[virtuals=fortran] gcc")
+
+
+# import spack_repo.builtin_mock.build_systems.compiler
+
+
+def _display_with_compilers(spec):
```

Comment:
This looks like a vestigial debug apparatus? If it's worth keeping, it's probably worth moving it somewhere like `spec.py`.